### PR TITLE
Update repo CMake configuration to target /W4 by default

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Contributes to https://github.com/dotnet/runtime/issues/66154 

Disable all failing warnings. The ones listed in the SDL issue above will gradually be re-enabled.

/cc @elinor-fung @am11 @jkotas

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -551,8 +551,8 @@ if (MSVC)
   # /W3 is added by default by CMake, so remove it
   string(REPLACE "/W3" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
   string(REPLACE "/W3" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
-  # set default warning level to 3 but allow targets to override it.
-  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/W$<GENEX_EVAL:$<IF:$<BOOL:$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>>,$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>,3>>>)
+  # set default warning level to 4 but allow targets to override it.
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/W$<GENEX_EVAL:$<IF:$<BOOL:$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>>,$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>,4>>>)
```
Comment (jkoritzinsky) [inline eng/native/configurecompiler.cmake:555]: If our minimum is W4 for the repo as a whole and we won’t add any components that use a lower warning level, we can probably simplify this to just `/W4`

Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:555]: Agreed. Once we have the entire repo able to do that, that is what will happen. However, I'd like to not impose that right now.

Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:555]: This can wait for later cleanup

File: eng/native/configurecompiler.cmake [hunk 2]
```diff
@@ -579,19 +579,37 @@ if (MSVC)
...
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4201>) # nonstandard extension used : nameless struct/union
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4206>) # nonstandard extension used : translation unit is empty
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4239>) # nonstandard extension used : 'token' : conversion from 'type' to 'type'
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4244>) # conversion from 'type1' to 'type2', possible loss of data
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4245>) # conversion from 'type1' to 'type2', signed/unsigned mismatch
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4291>) # no matching operator delete found; memory will not be freed if initialization throws an exception
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4310>) # cast truncates constant value
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4324>) # 'struct_name' : structure was padded due to __declspec(align())
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4366>) # The result of the unary 'operator' operator may be unaligned
```
Comment (elinor-fung) [inline eng/native/configurecompiler.cmake:593]: Looks like we also have some 4389 warnings.

File: eng/native/configurecompiler.cmake (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:555]: The native tests for instance will need to specify /W3.

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -551,8 +551,8 @@ if (MSVC)
   # /W3 is added by default by CMake, so remove it
   string(REPLACE "/W3" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
   string(REPLACE "/W3" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
-  # set default warning level to 3 but allow targets to override it.
-  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/W$<GENEX_EVAL:$<IF:$<BOOL:$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>>,$<TARGET_PROPERTY:MSVC_WARNING_LEVEL>,3>>>)
+  # set default warning level to 4 but allow targets to override it.
```
Comment (elinor-fung) [inline eng/native/configurecompiler.cmake:554]: The coreclr/jit should be able to stop overriding now, right?

File: eng/native/configurecompiler.cmake (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:593]: Those are only for the test projects. I'm not convinced we should fix those though. Thoughts?

Comment (elinor-fung) [reply eng/native/configurecompiler.cmake:593]: It is not just test projects:
```
D:\a\_work\1\s\src\coreclr\vm\object.cpp(582): warning C4389: '==': signed/unsigned mismatch
```

File: eng/native/configurecompiler.cmake (hunk 1 above)
Comment (elinor-fung) [inline eng/native/configurecompiler.cmake:555]: Do you want to add that `# [[! Microsoft.Security.SystemsADM.10086 !]]` marker while you are in here?

File: eng/native/configurecompiler.cmake (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:593]: @elinor-fung Which is that for? I didn't see that on Win x64.

File: eng/native/configurecompiler.cmake (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:554]: sure

File: eng/native/configurecompiler.cmake (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:555]: What is that?

File: eng/native/configurecompiler.cmake (hunk 2 above)
Comment (elinor-fung) [reply eng/native/configurecompiler.cmake:593]: It is on the windows arm leg:
https://dev.azure.com/dnceng/public/_build/results?buildId=1657065&view=logs&jobId=29c18220-1650-5afc-169a-2e7f66607d7e&j=29c18220-1650-5afc-169a-2e7f66607d7e&t=570d5850-cebf-50e8-1beb-95128f85b373

File: eng/native/configurecompiler.cmake (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply eng/native/configurecompiler.cmake:555]: Nevermind, I found it.

Comment (elinor-fung) [reply eng/native/configurecompiler.cmake:555]: https://github.com/dotnet/runtime/issues/66154 says to include `Microsoft.Security.SystemsADM.10086` in the file configuring the compilation to help with auditing. That specific string is from the example for adding it to a cmake file.

**Review by elinor-fung — APPROVED** [review]

File: src/tests/Interop/COM/NativeClients/Dispatch/Client.cpp [hunk 4]
```diff
@@ -364,7 +364,7 @@ void Validate_LCID_Marshaled()
         nullptr
     ));
 
-    THROW_FAIL_IF_FALSE(lcid == V_I4(&result));
+    THROW_FAIL_IF_FALSE(lcid == (LCID)V_IU4(&result));
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeClients/Dispatch/Client.cpp:367]:
```suggestion
    THROW_FAIL_IF_FALSE(lcid == (LCID)V_UI4(&result));
```

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**